azkaban-common/src/main/java/azkaban/user/UserUtils.java

```diff
@@ -100,38 +101,52 @@ static void setupWatch(final Map<String, ParseConfigFile> configFileMap) throws
       log.warn("Watchservice was not setup for any config file(s).");
       try {
         watchService.close();
-      } catch (IOException e) {
+      } catch (final IOException e) {
         log.warn("IOException while closing watchService. " + e.getMessage());
       }
       return;
     }
 
-    Runnable runnable = () -> {
+    final Runnable runnable = () -> {
       // Watchservice is established, now listen for the events till eternity!
-      for (;; ) {
-        WatchKey watchKey;
+      for (; ; ) {
+        final WatchKey watchKey;
         try {
           watchKey = watchService.take();
-        } catch (InterruptedException ie) {
+          // Wait for a second to ensure there is only one event for a modification.
+          // For a file update, WatchService creates two ENTRY_MODIFY events, 1 for content and 1
+          // for modification time.
+          // Adding the sleep consolidates both the events into one with a count of 2 which
+          // avoids multiple reloads of same file.
+          // One second seems excessive, however, these events happen very less often and it is
+          // more important that the config reloads successfully than immediately.
+          // If there is any modification happening to file(s) in the meantime, it is all queued up
+          // in the watch service.
+          Thread.sleep(1000L);
+        } catch (final InterruptedException ie) {
           log.warn(ie.getMessage());
           Thread.currentThread().interrupt();
           return;
         }
 
         // Get the directory for which watch service event triggered.
-        Path dir = keys.get(watchKey);
-        for (WatchEvent<?> event : watchKey.pollEvents()) {
+        final Path dir = keys.get(watchKey);
+        for (final WatchEvent<?> event : watchKey.pollEvents()) {
           // Make sure the modification happened to user config file
-          @SuppressWarnings("unchecked")
-          final Path name = ((WatchEvent<Path>) event).context();
+          @SuppressWarnings("unchecked") final Path name = ((WatchEvent<Path>) event).context();
           final String filename = dir.resolve(name).toString();
           // Lookup the file in dirToFilesMap
-          if (dirToFilesMap.containsEntry(dir, filename)) {
-            // Match!
-            // reparse the config file
-            log.info("Modification detected, reloading config file " + filename);
+          if (!dirToFilesMap.containsEntry(dir, filename)) {
+            continue;
+          }
+          // Match!
+          // Reparse the config file
+          log.info("Modification detected, reloading config file " + filename + ".");
+          try {
             configFileMap.get(filename).parseConfigFile();
-            break;
+          } catch (final Exception e) {
+            // If there is any exception while parsing the config file, log it and move on
+            log.warn("Reload failed for config file " + filename + " due to " + e.getMessage());
```

Comment:
I suppose just `+ e` ie. toString() may be usually nicer than `e.getMessage()` because message might be null.

Reply:
Good idea. Updating it everywhere in the patch.